packages/yew-router/src/navigator.rs

```diff
@@ -0,0 +1,176 @@
+        Self {
+            inner: history,
+            basename,
+        }
+    }
+
+    /// Returns basename of current navigator.
+    pub fn basename(&self) -> Option<&str> {
+        self.basename.as_deref()
+    }
+
+    /// Navigate back 1 page.
+    pub fn back(&self) {
+        self.go(-1);
+    }
+
+    /// Navigate forward 1 page.
+    pub fn forward(&self) {
+        self.go(1);
+    }
+
+    /// Navigate to a specific page with a `delta` relative to current page.
+    ///
+    /// See: <https://developer.mozilla.org/en-US/docs/Web/API/History/go>
+    pub fn go(&self, delta: isize) {
+        self.inner.go(delta);
+    }
+
+    /// Pushes a [`Routable`] entry.
+    pub fn push(&self, route: impl Routable) {
+        self.inner.push(self.route_to_url(route));
+    }
+
+    /// Replaces the current history entry with provided [`Routable`] and [`None`] state.
+    pub fn replace(&self, route: impl Routable) {
+        self.inner.replace(self.route_to_url(route));
+    }
+
+    /// Pushes a [`Routable`] entry with state.
+    pub fn push_with_state<T>(&self, route: impl Routable, state: T)
+    where
+        T: 'static,
+    {
+        self.inner.push_with_state(self.route_to_url(route), state);
+    }
+
+    /// Replaces the current history entry with provided [`Routable`] and state.
+    pub fn replace_with_state<T>(&self, route: impl Routable, state: T)
+    where
+        T: 'static,
+    {
+        self.inner
+            .replace_with_state(self.route_to_url(route), state);
+    }
+
+    /// Same as `.push()` but affix the queries to the end of the route.
+    pub fn push_with_query<Q>(&self, route: impl Routable, query: Q) -> NavigationResult<()>
+    where
+        Q: Serialize,
+    {
+        self.inner.push_with_query(self.route_to_url(route), query)
+    }
+
+    /// Same as `.replace()` but affix the queries to the end of the route.
+    pub fn replace_with_query<Q>(&self, route: impl Routable, query: Q) -> NavigationResult<()>
+    where
+        Q: Serialize,
+    {
+        self.inner
+            .replace_with_query(self.route_to_url(route), query)
+    }
+
+    /// Same as `.push_with_state()` but affix the queries to the end of the route.
+    pub fn push_with_query_and_state<Q, T>(
+        &self,
+        route: impl Routable,
+        query: Q,
+        state: T,
+    ) -> NavigationResult<()>
+    where
+        Q: Serialize,
+        T: 'static,
+    {
+        self.inner
+            .push_with_query_and_state(self.route_to_url(route), query, state)
+    }
+
+    /// Same as `.replace_with_state()` but affix the queries to the end of the route.
+    pub fn replace_with_query_and_state<Q, T>(
+        &self,
+        route: impl Routable,
+        query: Q,
+        state: T,
+    ) -> NavigationResult<()>
+    where
+        Q: Serialize,
+        T: 'static,
+    {
+        self.inner
+            .replace_with_query_and_state(self.route_to_url(route), query, state)
+    }
```

Comment:
Why duplicate the entire `History` trait from `gloo`?

Reply:
It does 2 things:

1. It shields `listen()`(Prevent an issue similar to #2116) and `location()` (make location a separate hook, so navigator hooks will become "setter-only" hence do not need to be updated when location changes.) method away from the user. 
2. It adds basename handling.

Reply:
Makes sense. I would've preferred to avoid duplication but that would cause fragmentation in the API which is arguably worse than what we have right now.